### ISSUE TITLE
Uncaught Throw not reported

### ISSUE THREAD

**Issue by szhorvat** [WolframResearch/WolframLanguageForJupyter#23]
A naked `Throw` should result in a `Throw::nocatch` message. There is currently no message, but the session does not break.

Command line (i.e. the expected behaviour):

```
In[1]:= Throw[123]                                                              

Throw::nocatch: Uncaught Throw[123] returned to top level.

Out[1]= Hold[Throw[123]]
```

Jupyter:

![image](https://user-images.githubusercontent.com/1212871/58006587-b2117280-7ae8-11e9-8c0b-e3a2a0f11f45.png)

The second cell illustrates that the session did not break (but the first cell produced no message).

Note: One is not going to evaluate such a Throw on purpose, but accidents happen, and they should be handled gracefully.

**Comment by szhorvat** [comment]
Correction: it _does_ break the session.

![image](https://user-images.githubusercontent.com/1212871/58007718-2c42f680-7aeb-11e9-939a-6c6266457933.png)

Another (perhaps more frivolous) way to break the session is to evaluate `Dialog[]`.

**Comment by szhorvat** [comment]
Given the large number of issues in the class "it doesn't work like a usual Mathematica REPL", would it not be better to use two kernels? The 1st kernel manages the connection to Jupyter. The 2nd kernel is a slave of the first kernel, managed through MathLink.  Only the 2nd kernel is used for computations.

This would have many advantages:

 - The user would not be able to mess up or even detect the Jupyter communication code that's running (try e.g. `Stack[]` which shows that we're not at the top level, or `$ContextPath` which unexpectedly has the ZMQ stuff loaded).

 - Things like `$Line`, `%`, etc. would work without hacks.

 - `Throw[]`, `Dialog[]`, `Return[]`, etc. would also just work.

 - If the computation kernel died, it could be detected and managed gracefully.

Disadvantages:

 - We would effectively have half as many licenses, as each session would use up two kernels. But this is an artificial problem, and I am sure that WRI could come up with a solution (either provide more licenses or run one kernel with a special license, just like the auxiliary kernel used by the front end).

It seems to me that this is really The Right Way to go. The kernel was designed to be controlled by front ends through MathLink, not ZMQ.

Alternatively, write the Jupyter glue code in another language than WL. It could be written in C++ ([similar things have been done before](https://github.com/QuantStack/xeus-cling) so you can re-use some bits), or it could be written in Python if MathLink were finally exposed to Python ...

**Comment by cc-wr** [comment]
> The 1st kernel manages the connection to Jupyter. The 2nd kernel is a slave of the first kernel, managed through MathLink. Only the 2nd kernel is used for computations.

Thank you for the suggestion, and it makes sense:

This has been discussed, and there is an internal branch for it. However, for simplicity's sake, evaluations are currently just evaluated in place in the script run by Jupyter.

Of course, this may, and probably will, change.

> Alternatively, write the Jupyter glue code in another language than WL. It could be written in C++ (similar things have been done before so you can re-use some bits), or it could be written in Python if MathLink were finally exposed to Python ...

It was an intentional design decision to have the project written natively in the Wolfram Language.

This does have various costs in certain places.

**Comment by szhorvat** [comment]
> This has been discussed, and there is an internal branch for it.

I wouldn't mind giving that a try, if possible :-)

**Comment by cc-wr** [comment]
Does this commit fix the problem to an adequate degree, for the moment, for you: 81ba1a6a6750aa1d304a88a345b1a0f71c3ddc98?

**Comment by szhorvat** [comment]
> Does this commit fix the problem to an adequate degree

I think so.

**Comment by szhorvat** [comment]
Sorry to write here, but I don't know how else to reach you. In case you weren't aware, there is a simple way to implement kernels in C++:

https://github.com/QuantStack/xeus
https://xeus.readthedocs.io/en/latest/kernel_implementation.html#implementing-the-interpreter

This may be a more robust way to create a WL Jupyter kernel that handles the WL Engine Kernel through MathLink.

**Comment by cc-wr** [comment]
It is no problem, but you have mentioned xeus before. (It was also discussed in the earlier stages of this project.)

Here, its benefits are that a lot of stuff is already implemented, and that some problems related to nature of the Wolfram Language can be more easily worked around (e.g., #25). So, I agree that it can, in certain aspects, be a path to a more robust Wolfram Language Jupyter kernel. 

However, on the MathLink issue, it is not like there are not easy MathLink bindings directly in the Wolfram Language. Personally, I see no reason to use xeus just for that.

**Comment by cc-wr** [comment]
This (#23) should be resolved.

**Comment by cc-wr** [comment]
Also, I am happy to continue the xeus discussion.

**Comment by szhorvat** [comment]
Sorry, I did not realize that I already suggested xeus by name. I thought I only suggested implementing the Jupyter communication in C++, but not a specific tool to do it.

The main point I was making was not really using xeus, but using a separate computation kernel that actually runs [Mathematica's main loop](http://reference.wolfram.com/language/tutorial/WolframSystemSessions.html). Many of the issues with the current version of this project seem to stem from not using the actual main loop but trying to emulate it. Maybe I'm wrong though, I must admit I did not dig very deep into your code after the internal version was merged.

In theory, one advantage of xeus is that we'd use only one kernel license (not two: one for the manager kernel and one for the compute kernel). I guess WRI could change how licensing works to get around this, if they wanted to.  I'll admit that I did not think through the disadvantages of xeus (e.g. how do we handle image outputs, rasterize 2D output, etc.—generally make the interface practically usable and be more than a fancy command line)
